SG-3055: Stand alone publish API

josh-t opened the pull request:
The UI portion of this will take some time, so I'm submitting the API in its current state for initial review. I realize there is some work to do regarding docstrings but for the most part the code is in a state where it needs new eyes. 

There code is organized in a way very similar to the existing code in `processing`. The biggest difference is the externalized representation of the graph itself in `graph.py`. I would suggest, for review, starting with `manager.py` and then looking at `graph.py` to see if the organization seems reasonable. 

The `PublishManager` provides the interface to publishing and stores a `PublishGraph` instance. My hope is that having the graph representation as its own thing will make creating a custom item model in Qt easier, but I'm not sure. The graph is probably where I have the least confidence in the current state of things, so new eyes there would be great.

In order to allow existing publish plugins to work with the new API, `PublishItem` and `PublishTask` instances keep a reference to the graph. Would be good to get input here as well.

The `PublishData` class serves as a simple base class for both item properties and settings. This was to aid in serialization when I was expecting to have to do things more manually. 

Let me know if you have questions. Adding @manneohrstrom and @jfboismenu since it's a good chunk of code.

josh-t commented on python/tk_multi_publish2/api/data.py:
```diff
@@ -0,0 +1,58 @@
+# Copyright (c) 2017 Shotgun Software Inc.
```
This existed somewhat before here: https://github.com/shotgunsoftware/tk-multi-publish2/blob/master/python/tk_multi_publish2/processing/item.py#L646

It's now the base class for item properties and for hook/plugin settings.

josh-t commented on python/tk_multi_publish2/api/graph.py:
```diff
@@ -0,0 +1,521 @@
+# Copyright (c) 2018 Shotgun Software Inc.
```
This is all new.

josh-t commented on python/tk_multi_publish2/api/item.py:
```diff
@@ -0,0 +1,416 @@
+# Copyright (c) 2018 Shotgun Software Inc.
```
This existed before here: https://github.com/shotgunsoftware/tk-multi-publish2/blob/master/python/tk_multi_publish2/processing/item.py

I tried to strip it down to eliminate stuff that was UI-specific. The interface has to be the same, but some of the implementations have changed. Still work to do here as you can see at the end of the file.

josh-t commented on python/tk_multi_publish2/api/manager.py:
```diff
@@ -0,0 +1,446 @@
+# Copyright (c) 2018 Shotgun Software Inc.
```
This object inherits much of what https://github.com/shotgunsoftware/tk-multi-publish2/blob/master/python/tk_multi_publish2/processing/plugin_manager.py did. The implementation is different though.

josh-t commented on python/tk_multi_publish2/api/plugins/collector_instance.py:
```diff
@@ -0,0 +1,88 @@
+# Copyright (c) 2017 Shotgun Software Inc.
```
I believe this is identical to https://github.com/shotgunsoftware/tk-multi-publish2/blob/master/python/tk_multi_publish2/processing/plugin.py#L410-L478

josh-t commented on python/tk_multi_publish2/api/plugins/instance_base.py:
```diff
@@ -0,0 +1,157 @@
+# Copyright (c) 2017 Shotgun Software Inc.
```
Pretty close to https://github.com/shotgunsoftware/tk-multi-publish2/blob/master/python/tk_multi_publish2/processing/plugin.py#L19 with the added `id` property and the `__getstate__`, `__setstate__` methods (which I haven't documented yet... I'll do this asap).

josh-t commented on python/tk_multi_publish2/api/plugins/publish_plugin_instance.py:
```diff
@@ -0,0 +1,211 @@
+# Copyright (c) 2018 Shotgun Software Inc.
```
Basically same as this: https://github.com/shotgunsoftware/tk-multi-publish2/blob/master/python/tk_multi_publish2/processing/plugin.py#L118 with the added get/set state methods used for pickling.

josh-t commented on python/tk_multi_publish2/api/plugins/setting.py:
```diff
@@ -0,0 +1,39 @@
+# Copyright (c) 2017 Shotgun Software Inc.
```
New version of this: https://github.com/shotgunsoftware/tk-multi-publish2/blob/master/python/tk_multi_publish2/processing/setting.py#L16

josh-t commented on python/tk_multi_publish2/api/task.py:
```diff
@@ -0,0 +1,147 @@
+# Copyright (c) 2018 Shotgun Software Inc.
```
New version of this: https://github.com/shotgunsoftware/tk-multi-publish2/blob/master/python/tk_multi_publish2/processing/task.py

Internals rewired a bit.

josh-t commented on python/tk_multi_publish2/api/graph.py:
```diff
@@ -0,0 +1,588 @@
+        return graph_str
+
+    def _remove_items(self, items):
+        """
```
This is the method that makes me wonder if this is the right approach. The internal state of the object is super simple and the lookups are really easy to manage when it comes to traversing, adding items, and querying relationships, but removing items is a bit complex.

jfboismenu replied:
Storing the dependencies as a three when persisting will be fine I think.

As for ids, you don't need to build them all up front. You can probably use the Python object's id as the model id. Or you can create your own mapping of id's and weakrefs to the item/task.

manneohrstrom replied:
I think we let worries about serialization be an argument for a less DX-friendly API. Serialization and the design of it can help motivate a certain approach, but i definitely think we should start by looking at "what do we want the TD experience to be like" and then figure out what the internals look like.

I had similar thoughts to JFs', but manage to lose my comments as a accidentally cancelled my CR (github was a little bit confusing if you are trying to dismiss the submit UI for CR!)

josh-t commented on python/tk_multi_publish2/api/item.py:
```diff
@@ -0,0 +1,423 @@
+        return self._graph.get_parent(self)
+
+    @property
+    def persistent(self):
```
This is a new property for items that should not be removed between collections (like dragged/dropped files). It's a more generic approach than what we did manually in the pre existing code. This also allows clients who build/manage their graph can mark any items as persistent if they wish.

josh-t commented on python/tk_multi_publish2/api/item.py:
```diff
@@ -0,0 +1,423 @@
+                "Could not determine the current publish plugin when accessing "
+                "an item's local properties. Item: %s" % (self,))
+
+        plugin_id = calling_object.id
```
We used to cache the object itself as the key to the lookup. The id property was added to plugins and used here so that items could be serialized/deserialized.

josh-t commented on python/tk_multi_publish2/api/manager.py:
```diff
@@ -0,0 +1,444 @@
+
+        self.publish_logger.info("Executing publish pass...")
+
+        # TODO: re-validate?
```
Not sure if validation should happen again here in the API. We do this on the UI side, but maybe for the UI we don't want to enforce a validate on every publish?

jfboismenu replied:
I think it's fine to publish without doing validation in the API. Not everything needs to be validated all the time when using the API directly. Validation is there for a end-user. Devs are fine with an exception being raised I think.

manneohrstrom replied:
yes, validation should be a "TD-opt in" - we shouldn't force all workflows to be validated - but a TD should be able to enforce validation of input. 

If a user is writing a custom UI that replaces the publish UI, they can still run the validation programatically, right?

jfboismenu commented on python/tk_multi_publish2/api/item.py:
```diff
@@ -0,0 +1,435 @@
+        """
+
+        # try to determine the current publish plugin
+        calling_object = inspect.stack()[2][0].f_locals.get("self")
```
Have you considered simply walking up the stack? This seems way too restrictive.

jfboismenu commented on python/tk_multi_publish2/api/item.py:
```diff
@@ -0,0 +1,435 @@
+        plugin_id = calling_object.id
+
+        if plugin_id not in self._local_properties:
+            self._local_properties[plugin_id] = PublishData()
```
Maybe use a `collections.defaultdict` for `_local_properties`?

josh-t replied:
ah right! done. 😛

jfboismenu commented on python/tk_multi_publish2/api/item.py:
```diff
@@ -0,0 +1,435 @@
+        return self._local_properties[plugin_id]
+
+    # TODO: consider
+    def set_icon_from_path(self, path):
```
What's the difference between icon and thumbnail? Also, why not properties?

josh-t commented on python/tk_multi_publish2/api/manager.py:
```diff
@@ -0,0 +1,446 @@
+                )
+
+            # retrieve the settings from the returned list of dictionaries
+            app_settings = context_settings[0]["settings"]
```
good call. fixed.

josh-t commented on python/tk_multi_publish2/api/plugins/instance_base.py:
```diff
@@ -0,0 +1,157 @@
+        self._logger = publish_logger
+
+        # every plugin created for use in the graph has a unique id
+        self._id = uuid.uuid4().hex
```
having used this to debug, i don't think it's worth the effort to change. it's not bad really.

josh-t commented on python/tk_multi_publish2/api/plugins/instance_base.py:
```diff
@@ -0,0 +1,157 @@
+        try:
+            hook_settings_schema = self._hook_instance.settings
+        except AttributeError, e:
+            import traceback
```
left over from debugging. removed.

jfboismenu commented on python/tk_multi_publish2/api/graph.py:
```diff
@@ -0,0 +1,591 @@
+import traceback
+
+try:
+    import cPickle as pickle
```
Yeah, my comment was the dual imports. I don't know why cPickle might be missing either.

jfboismenu commented on python/tk_multi_publish2/api/graph.py:
```diff
@@ -0,0 +1,591 @@
+    """
+
+    @staticmethod
+    def load(file_path):
```
Then maybe the only the properties dictionary needs to be pickled in that case, since we can't control it's content.

jfboismenu commented on python/tk_multi_publish2/api/graph.py:
```diff
@@ -0,0 +1,591 @@
+                yield child_item
+
+
+class _PublishGraphUnpicklerFactory(object):
```
At the very least, we can certainly investigate it as a SPIKE at some point.

jfboismenu commented on python/tk_multi_publish2/api/graph.py:
```diff
@@ -0,0 +1,591 @@
+
+        # given a namespaced module/classname, return the same module in the new
+        # tk namespace. example module name:
+        #     tkimp4b0b5e4c49974e21813a21928b101a9f.tk_multi_publish2.base_hooks
```
Maybe it's just because I'm not super familiar with pickle, but how would the unpickler deal with a class rename or a class broken up in two?

At least with JSON we control the entire serialization/deserialization process. Also, JSON allows offline editing of the file, since it isn't tied to any Python logic. That might be useful for TDs who want to patch something on a series of file.

manneohrstrom commented:
Maybe i am overthinking this, but would it make sense - as we look at the API changes JF is proposing - to create some example scripts to show the use of the API? I am thinking one example showing how the API can be used to have a custom UI drive the publish process. And a second example(s) where the publisher is using its UI but serializes to file and this is picked up and run on the compute farm (could we even write out the record to Shotgun?)

I found in the past that sometimes it's useful to "code review" the DX/usage patterns as much as the code itself - and often as you start putting together the samples, you realise patterns and not obvious just by looking at the code!

These samples could also be used for the documentation later on, so it's not wasted work!

josh-t commented:
@jfboismenu @manneohrstrom 
Hey guys, this is ready for another round. I think I've addressed the concerns JF raised. The API now acts like the existing code whereby items own their children. The "tree" instance (formerly "graph") provides the API for acting on the items as a whole (iteration, serialization, etc). The tree itself isn't exposed to the user directly. Rather, the "manager" API gives us a way to expose high-level publish functionality (publish, validate, save, load, etc) and selectively expose more granular methods/properties as needed. Have another look and see how it feels.

Here is some example code for either end of a collect+serialize+save -> load+deserialize+publish workflow:

```python
# get the app instance and create a publish manager
import sgtk
engine = sgtk.platform.current_engine()
publish_app = engine.apps["tk-multi-publish2"]
manager = publish_app.create_publish_manager()

# collect items for the current session and manually add an image to publish. this will handle item 
# creation and task assignment based on the destination context.
manager.collect_session()
manager.collect_files(["/Users/tomlinj/Documents/load.png"])

# save the configured publish to disk
manager.save("/tmp/example.pub")
``` 
Here's an example of how the json serialization currently looks on disk:
[example.json.txt](https://github.com/shotgunsoftware/tk-multi-publish2/files/2190798/example.json.txt)
```python
# same as ^^^
import sgtk
engine = sgtk.platform.current_engine()
publish_app = engine.apps["tk-multi-publish2"]
manager2 = publish_app.create_publish_manager()

# load the serialized publish from disk and publish it
manager2.load("/tmp/example.pub")
manager2.publish()
```
I'm sure there will be bugs/issues to work out, plus tests and docs to write as we get into the UI conversion but this feels a lot cleaner and maintainable. Please have a look!

josh-t commented:
Regarding the other cases you bring up @manneohrstrom, here's what I was thinking...

----
For the case where you're having artists run the validation in the UI, but you're executing the actual publish elsewhere, you'd need to identify what needs to be processed locally and what doesn't as the code is executing. I can imagine clients will want to be able to do the execution deferral at both the item level (defer all tasks) and the individual task level (do the expensive stuff on the farm for this item). You still want to be able to modify the settings via custom UI on the tasks regardless of when they execute, so all the existing code is still applicable. 

For item-level deferral, you could set a property during collection:
```python
item.properties.deferred = True
```
Plugins would then need to check for this property and decide whether they should execute or not (do I have a UI? am I on a "farm" machine, or whatever the logic is). This would likely need to be done in both the `publish()` and `finalize()` method since they always execute.

For task-based deferral, I imagine similar logic but without the need to check item properties. In other words, the plugins are written in a way that they know whether they should execute or not based on the current machine or the engine or whatever. 

In order to trigger the serialize, right now the client would need to do this in a `finalize()` somewhere. We'll likely need to expose the `manager` instance being used by the publisher somehow so that the client code has access to the `save` (serialization) at a minimum. We have an existing request for a post-publish that runs after the tree has been processed. Maybe we add a hook for that with the `manager` as an argument. 

I'm starting to think we should find a client to partner with to run through these scenarios (if possible).

----

For the custom UI scenario, a `manager` instance should be what is used. We will be able to validate this interface when we replace the existing UI's internals with the new code. I think it should use the `PublishManager` interface exclusively. If we can make that work, I'll feel pretty comfortable about clients doing the same. I can already think of a couple of ways to improve it, but this feels right to me. 

Having said that, it might make sense to provide the publish tree Qt model to clients as well. I'm not exactly sure how this will all fit together yet. This should become clearer when we dive into the UI side of the work.

manneohrstrom commented on app.py:
```diff
@@ -102,6 +108,17 @@ def context_change_allowed(self):
+        :param publish_logger: The logger to be used during publish processing.
+        :returns: A :class:`tk_multi_publish2.PublishManager` instance
+        """
+        return self._manager_class(publish_logger=publish_logger)
```
not sure this is the right idiom for logging - i think rather than this, we want a fixed logger to always do the logging and then we want to be able to attach a handler we create to that logger. What happens to logging to disk if you bring your own logger for example? 

I think the right way to do this is to be able to get the "address" of the publish logger so that you can then attach a handler to that logger. But we don't want users to provide their own logger.

josh-t replied:
Makes sense. I've gone ahead and removed this for now. I think things will clarify once we plug this into, and rework the UI since that will be a client of this API. If we match the existing behavior with respect to logging, I think it'll be clear what the right interface is.

josh-t commented:
@manneohrstrom yes, right now it's serializing everything it needs to re-instantiate plugins. Right now this includes the default values for configured plugins. To remove these, we'll need to update the deserialization logic to do `sgtk.platform.engine.find_app_settings()` to find/load these values since they're used as fallbacks during publish.

Regarding encoding the bootstrap settings, that's interesting. Maybe we could encode some information into the root item's properties and provide an interface for clients to validate against the current context. There may be valid reasons though to "templatize" a publish tree though. I think we should discuss more.

jfboismenu commented on python/tk_multi_publish2/api/manager.py:
```diff
@@ -0,0 +1,487 @@
+        Publish the collected items.
+
+        Each task assigned to collected items will execute their publish
+        payload. Items are processed in depth first order.
```
Maybe add a comment that makes it clear validate should be called first?

josh-t replied:
done.

jfboismenu commented on python/tk_multi_publish2/api/plugins/instance_base.py:
```diff
@@ -0,0 +1,140 @@
+    publish plugins).
+
+    Each object reflects an instance in the app configuration.
+    """
```
Could you add a serialized example?

josh-t replied:
same comment about making the serialized state public.

jfboismenu commented on python/tk_multi_publish2/api/task.py:
```diff
@@ -0,0 +1,174 @@
+    A task is a particular unit of work which can to be carried by the
+    publishing process. A task can be thought of as a 'plugin instance', or a
+    particular publishing plugin operating on a collected item.
+    """
```
An example of the serialized version would be nice.

josh-t replied:
same comment about making the serialized state public.

jfboismenu commented on python/tk_multi_publish2/api/task.py:
```diff
@@ -0,0 +1,174 @@
+                setting["default_value"],
+                setting["description"]
+            )
+            new_setting.value = setting["value"]
```
Didn't you point out there other day that the value could be anything?

Maybe we could so something where we serialize builtin types like collections, numbers and strings as is in the document and pickle everything else?

jfboismenu replied:
So the json would contain value or pickled_value depending on how it was serialized?

jfboismenu replied:
Keep in mind that when serializing collections, you need to recurse into them to know if there are non easily serializable objects.

I just realized that stuff like sets would need to be pickled since turning them into an array loses their semantic.

Maybe we should just always pickled everything? In any case, let's use pickled_value in the json document, just in case in the future we need to store unpickled versions of certain values.

josh-t replied:
These are the "settings" which have to be representable via the config anyway. So I think we're ok here. But your points are valid for the local/global properties on the item. Let me have another look to see what we should do here. I'm assuming this will be a major version release, so maybe we should just add a restriction as to what can be shoved in there.

jfboismenu replied:
Sorry, I got the two confused, you're totally right. I meant the global/local properties.

jfboismenu commented on python/tk_multi_publish2/api/tree.py:
```diff
@@ -0,0 +1,227 @@
+    Implements the json encoder interface for custom publish tree serialization.
+    """
+    def default(self, publish_tree):
+        return publish_tree.to_dict()
```
How about we serialize the tree with a schema number to track versions of this?

josh-t replied:
yeah, i meant to do something like this but forgot. I'll add something in there.

jfboismenu commented on python/tk_multi_publish2/api/tree.py:
```diff
@@ -0,0 +1,227 @@
+    Implements the json decoder interface for custom publish tree
+    deserialization.
+    """
+    def decode(self, json_str):
```
When decoding, you read that number and you pass it as a parameter to from_dict so that more recent versions of the code know which version of the schema the document is?

jfboismenu replied:
`def from_dict(schema_version=LATEST)` which is defined to whatever is the latest version.

josh-t commented on python/tk_multi_publish2/api/data.py:
```diff
@@ -0,0 +1,55 @@
+    """
+
+    @classmethod
+    def clone(cls, pub_data_obj):
```
I've always preferred classmethods for this type of thing, but maybe an instance method is more idiomatic. I'll do some searching to see what's more preferable.

jfboismenu replied:
Maybe it's because I'm coming at this from a C++ background.

Having a method implemented in each class called clone() that returns a pointer to the cloned object is pretty standard fare. The only difference here is that in Python we can get away with writing a single method, which is pretty nice.

jfboismenu replied:
besides

```
item = item.clone() 
```

feels nicer to me than

```
item = PublishData.clone(item)
```

jfboismenu replied:
Come to thing of if, shouldn't we have implemented this as
return copy.deepcopy(self)?

josh-t commented on python/tk_multi_publish2/api/item.py:
```diff
@@ -0,0 +1,531 @@
+    """
+    Items represent a "thing" to be operated on. This can be a file on disk
+    or an object of some type in an artist's session.
+    """
```
Do we want to make the serialized state part of the public docs? I'd prefer to keep it hidden so that we can make changes as needed.

jfboismenu replied:
Fair point. While the format allows to edit directly the file, we don't want to encourage that behaviour either. :)

Maybe just add it as a comment inside to_dict?

josh-t commented on python/tk_multi_publish2/api/item.py:
```diff
@@ -0,0 +1,531 @@
+        # assigned to the item, it will inherit it from the parent or current
+        # bundle context on deserialize.
+        if self._context:
+            context_value = {
```
Yeah, I think it would be pretty useful in general to have a way to get a context as a dict via the core API. I've created a ticket for it and added a TODO here.

josh-t commented on python/tk_multi_publish2/api/item.py:
```diff
@@ -0,0 +1,531 @@
+
+    @property
+    def children(self):
+        """Return the children of this item."""
```
I don't have a strong opinion. What do you think?

jfboismenu replied:
Well, if people are generally just going to iterate over them, `yield` has the benefit of never running the risk of having people of removing a children from the array instead of invoking our `remove_child` method.

jfboismenu replied:
It's not something we've been careful of in the past, but I would love if we could try to be more careful about in the future.

Especially since in Python 3 puts a heavy emphasis on iterators.

josh-t commented on python/tk_multi_publish2/api/item.py:
```diff
@@ -0,0 +1,531 @@
+        If this is called more than one level deep in this class, expect issues.
+        """
+
+        # TODO: walk up the stack to see if this is called by a hook
```
added.

josh-t commented:
I'm closing this one and leaving this branch parked. There's still plenty of work to do here with the other internal tickets. I've added comments, addressed JF's points and left additional TODOs to consider when this gets picked back up.
